--- /tmp/nsa_v018/codegen/v016/case06.host.cpp	2026-09-24 10:59:18.331728779 +0000
+++ /tmp/nsa_v018/codegen/v_during_softmax/case06.host.cpp	2026-09-24 10:58:38.374629898 +0000
@@ -2611,7 +2611,7 @@
   (((TVMFFIAny*)stack_ffi_any)[10].v_int64) = ((int64_t)1);
   (((TVMFFIAny*)stack_ffi_any)[11].type_index) = 1;
   (((TVMFFIAny*)stack_ffi_any)[11].zero_padding) = 0;
-  (((TVMFFIAny*)stack_ffi_any)[11].v_int64) = ((int64_t)8192);
+  (((TVMFFIAny*)stack_ffi_any)[11].v_int64) = ((int64_t)8448);
   (((TVMFFIAny*)stack_ffi_any)[12].type_index) = 0;
   (((TVMFFIAny*)stack_ffi_any)[12].zero_padding) = 0;
   (((TVMFFIAny*)stack_ffi_any)[12].v_int64) = (int64_t)0;
